design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,324 @@
+  that the feature is planned to be removed.
+  Warnings should be raised when:
+  - Feature API fields are accessed.
+  - Feature FG is activated.
+  - Behavior related to the feature is detected (optional).
+- Removal: Feature removal involves removing the core functionality
+  of a feature and its exposed API.
+  - The core implementation can be removed in two steps:
+    - The FG is removed by assuring it is never reported as set
+      (i.e. even if it is left by the operator configured, internally
+       it is ignored).
+      At this stage, the core implementation will follow the FG conditions
+      and therefore from the outside the feature is inactive.
+    - In case there are no side effects, the core implementation code can
+      be removed.
+  - The API types are not to be removed, as it may have implications
+    with the underlying storage which has already persisted them.
+    Kubernetes has not removed fields, it just kept them around with the
+    warning that they have been deprecated and no longer available.
+
+    While keeping fields around for a period of a release or two makes
+    sense, beyond a limited period it adds a burden on dragging leftover
+    fields around to eternity.
+
+> **Note**: The only reference seen on why fields should not be removed
+> was mentioned [here](https://github.com/kubernetes/kubernetes/issues/52185).
+> But it is unclear if this is relevant for Alpha stage features.
+> Starting with a strict policy, similar to Kubernetes is recommended,
+> i.e. once fields are introduced, they should not be removed no matter
+> the feature release stage.
+> Per need, the topic can be revisited in follow-up adjustments.
+
+### Exceptions
+While the project strives to maintain a stable contract with its users,
+there may be scenarios where the policy described here will not be a fit.
+
+Therefore, it should be acceptable to have exceptions from time to time
+given a very good reasoning and an agreement from 2/3 of the project
+maintainers (also known as "approvers").
+
+## Implementation Phases
+- Add a section in the
+  [design proposal template](https://github.com/kubevirt/community/blob/main/design-proposals/proposal-template.md)
+  that describes the planned timelines for the feature stages.
+  Add a reference to the feature-lifecycle documentation to assure contributors
+  know the process and policy.
+- Prepare a user-facing document that describes the usability implications of
+  this feature lifecycle.
+
+## Miscellaneous
+
+- New features are to be introduced to major and minor release versions only.
+- CI:
+  - Alpha stage features should not be gating on CI.
+  - Beta and GA features should be gating on CI.
+- API fields may be marked with the following information:
+  - Description
+  - Release stage (alpha/beta/ga) and release version.
+
+  It is left to a follow-up implementation proposal to define the exact format and
+  required/optional information.
```

Comment:
This last point is important and should probably be tackled as part of this proposal IMO.
It is crucial to have a place where we can easily see what is alpha vs beta.

Reply:
I wanted to defer most "how" details to a follow up in order not to block this process oriented proposal on such details (which have the potential to raise debate).

But if there is consensus and everyone is in agreement on how to do it, I am ok to add it.

Do you have a suggestion? I know we already have this added at some fields already.

Reply:
Ah, sorry I misunderstood, things get a bit meta when reviewing a proposal on how to write a document on how we'll write code :)
What you mean is "the document will describe a way to specify what is alpha/beta, this proposal just doesn't care how". I read "the document won't describe it, we'll figure that out later".
Maybe this can be clarified a bit to reflect precise intent, but just a nit at this point, feel free to leave as is.

Reply:
Well, I'm trying to push this to completion anyway possible.
In other words, tell me what is needed to get the lgtm :)

Reply:
What do we need to resolve this conversation?